CodingStandard.md

```diff
@@ -0,0 +1,1425 @@
+
+Note that the use of constructs or mechanisms for which justification is
+required is not discouraged. Requiring justification in these cases is simply
+intended to prevent misuse and to capture rationale for future reference.
+
+
+## General Standards
+
+**Language Specification**
+
+All production code should be written in the C++ programming language and should
+be compliant with the C++14 (ISO/IEC 14882:2014) language standard, except where
+such standards are not supported by the Robot Operating System (ROS) framework.
+If such instances are found, notify the Software Director so that they may be
+included in this document.
+
+**Coding Standard**
+
+Project standards and guidelines defined in this document should be adhered to
+for all production code. This document shall be maintained and updated as those
+standards are modified or replaced.
+
+**Names**
+
+Names (types, variables, functions, constants, namespaces, enumerators,
+macros, files, etc.) must be descriptive of their purpose and consistent.
+
+**Comments**
+
+Comments must be used to clarify parts of code that merit explanation,
+must be intelligible, and must be maintained.
+
+*[Question: Should we add a standard here for including Doxygen as a supported 
+
+tool?]*
+
+**Compilation**
+
+Code should compile and link with no errors or warnings at the `-Wall` `-Wextra`
+warning level.  Unavoidable warnings may be disabled locally, in as small a
+section of code as possible, with rationale stated in a comment wherever the
+warning is disabled.
+
+
+**#Includes**
+
+System-wide include files (e.g. stdafx.h) must contain a minimal amount of
+definitions.
+
+#include directives in a file must only be preceded by other preprocessor
+directives or comments. Exception: files that contain large data tables
+(e.g. a CRC coefficient lookup table) may be included at any location to
+promote readability.
+
+
+**Header Files**
+
+Every header file should be able to be compiled in isolation, meaning it
+includes any headers that its contents depend upon.  Relying on transitive
+#includes of other headers can make maintenance difficult and introduce
+subtle issues.
+
+
+**Memory**
+
+A function must not return a pointer or reference to memory allocated
+within that function with temporary or automatic storage duration.
+
+Classes that allocate memory with dynamic storage duration, including
+any memory allocated with the new operator, or allocate other shared resources
+must explicitly define copy constructors and copy assignment operators.
+
+
+**Internal/External Linkage**
+
+Variables with static storage duration that are not intended to be
+referenced outside of their compilation unit must be declared with internal
+linkage.
+
+Variables (and functions and types) declared and used within a single
+compilation unit should be declared inside an anonymous namespace, instead of
+using the `static` keyword.
+
+**Casting**
+
+C++ style casts must be used. Other cast formats (such as C-style
```

Comment:
Is this static_cast and dynamic_cast only?

Reply:
I'm not certain what you mean.  

A C-style cast, such as `(int)myVar` attempts to perform several different types of casts before reporting a failure (I believe it does a `static_cast` first, then a `dynamic_cast`, then a `reinterpret_cast`, and only if all three fail does it throw an exception).  Requiring c++ casts decreases the chance of a bug due to an incorrectly programmed cast.

Reply:
I wonder if Collin is talking about explicitly limiting the casts we are allowed to use to `static_cast` and `dynamic_cast`?

Reply:
Yeah I was I talking about limiting, to static_cast, but in general not subverting the type system. e.g. I don't want to see 
char c = inputMethod();
if ((int)char == 32;)
  ...

At least outside of a self contained utility method that does the conversion,.

Reply:
Technically, that would resolve to a `static_cast` :)